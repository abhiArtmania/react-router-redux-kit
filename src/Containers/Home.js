import React, { Component, useState, useEffect } from 'react';
import { connect } from 'react-redux'
import Header from './Header'
import { fetchData, fetchLanguages } from '../Actions/home'
import { bindActionCreators } from 'redux'
import { regex } from '../Constants'

const mapStateToProps = (state) => {
  return {
    homeInfo:state.home
  }
}

const mapDispatchToProps = dispatch => {
  return bindActionCreators(
    {
      fetchData,
      fetchLanguages
    },
    dispatch
  )
}


const Home = (props) => {
  const {history = {}} = props;
  const[state,setState] = useState({
    email:''
  })
  const[errors,setErrors]= useState({error:false})
  const { homeInfo, fetchData, fetchLanguages } = props
  const userInfo = homeInfo.user
  useEffect(() => {
    fetchData()
  }, [])

  const submitEmail = () => {
    if (!state.email || !regex.email.test(state.email)){
      setErrors( {error:'Please enter a valid email!'} );
    } else {
      fetchLanguages(()=>{
        setErrors( {error:false} );
      });
    }
  }
  return (
    <div className="home">
      <Header/>
      <div className='wrapper'>
        <div><h3 data-testid="home-txt">Home Page</h3></div>
        <p>
          <span>
          <div>{`Name: ${userInfo.name}`}</div>
          <div>{`Location: ${userInfo.location}`}</div>
          <div>{`Email: ${state.email}`}</div>
          </span>
        </p>
        <span>
          <input
            data-testid='email-input'
            name='email'
            type='text'
            value={state.email}
            onChange={(e)=>{
              setState({...state,['email']:e.target.value})
            }}
          />
          <button data-testid='button' onClick={submitEmail}>Get language list</button>
          {errors.error && <div data-testid='error-message' style={{color:'red'}}>{errors.error}</div>}
        </span>
        <div>
          <p>
            <span data-testid='about-us-link' style={{cursor:'pointer'}} onClick={()=>{
              history.push('/about')
            }}>About Us</span>
          </p>
        </div>
      </div>
      {homeInfo && homeInfo.languages && homeInfo.languages.length ? <div data-testid="languages-list">
        <h3>Available Language</h3>
        <ul>
          {homeInfo.languages.map((l,i)=>{
            return <li key={i}>{l.name}</li>
          })}
        </ul>
      </div> : ''}
    </div>
  );
}

export default connect(mapStateToProps, mapDispatchToProps)(Home)
