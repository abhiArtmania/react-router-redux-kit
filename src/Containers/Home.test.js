import React from 'react';
import { shallow, mount } from 'enzyme';
import Home from './Home';
import configureMockStore from "redux-mock-store";
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

const mockStore = configureMockStore();

let wrapper;
let store;
store = mockStore({
  home:{
    user:{
      name:'Abhishek',
      location:'Varanasi'
    }
  }
})
const historyMock = { push: jest.fn() };

beforeEach(() => {
  wrapper = mount(<Provider store={store} key="provider">
      <Router>
        <Home history={historyMock} />
      </Router>
    </Provider>);
});

afterEach(() => {
  jest.resetAllMocks();
});

describe("Home page", () => {
  it("should render properly", () => {
      expect(wrapper).toMatchSnapshot();
  })
})

describe("Home page test",()=>{
  it("Header renders properly in Home page",()=>{
    expect(wrapper.find('Header').exists()).toBe(true)
  })

  it("should have header - Home page",()=>{
    expect(wrapper.exists('[data-testid="home-txt"]')).toBe(true);
  })

  it("Should have input for email",()=>{
    expect(wrapper.find('[data-testid="email-input"]').length).toEqual(1)
  })

  it("Should have a button",()=>{
    expect(wrapper.find('[data-testid="button"]').length).toStrictEqual(1)
  })

  it('should respond to change event and change the state of the Home Component for email', () => {
    wrapper.find('[data-testid="email-input"]').hostNodes().simulate('change', { target: { name: 'email', value: 'test@gmail.com' } });
    expect(wrapper.find('[data-testid="email-input"]').hostNodes().props().value).toEqual('test@gmail.com');
  })

  it('should invoke history push on click of About us link', () => {
    wrapper.find('[data-testid="about-us-link"]').hostNodes().simulate('click');
    expect(historyMock.push).toHaveBeenCalledTimes(1);
    expect(historyMock.push.mock.calls[0]).toEqual(['/about']);
  })

  it("Error should be displayed when email is invalid and form is submitted",()=>{
    wrapper.find('[data-testid="email-input"]').hostNodes().simulate('change', { target: { name: 'email', value: '' } });
    const emailValue = wrapper.find('[data-testid="email-input"]').hostNodes().props().value
    wrapper.find('[data-testid="button"]').hostNodes().simulate('click');
    expect(wrapper.find('[data-testid="error-message"]').length).toEqual(1)
  })

})