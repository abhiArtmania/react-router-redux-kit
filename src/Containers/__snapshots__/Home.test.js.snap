// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page should render properly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(Home)
        history={
          Object {
            "push": [MockFunction],
          }
        }
      >
        <Home
          fetchData={[Function]}
          fetchLanguages={[Function]}
          history={
            Object {
              "push": [MockFunction],
            }
          }
          homeInfo={
            Object {
              "user": Object {
                "location": "Varanasi",
                "name": "Abhishek",
              },
            }
          }
        >
          <div
            className="home"
          >
            <Header>
              <div
                className="app-header"
              >
                <header
                  className="App-header"
                >
                  <img
                    alt="logo"
                    className="App-logo"
                    src="logo.svg"
                  />
                  <h4
                    className="App-title"
                  >
                    Welcome to React
                  </h4>
                  <MainMenu>
                    <div
                      className="main-menu"
                    >
                      <Link
                        replace={false}
                        to="/"
                      >
                        <a
                          href="/"
                          onClick={[Function]}
                        >
                          <button>
                            Home
                          </button>
                        </a>
                      </Link>
                      <Link
                        replace={false}
                        to="/about"
                      >
                        <a
                          href="/about"
                          onClick={[Function]}
                        >
                          <button>
                            About
                          </button>
                        </a>
                      </Link>
                      <Link
                        replace={false}
                        to="/services"
                      >
                        <a
                          href="/services"
                          onClick={[Function]}
                        >
                          <button>
                            Services
                          </button>
                        </a>
                      </Link>
                      <Link
                        replace={false}
                        to="/contact"
                      >
                        <a
                          href="/contact"
                          onClick={[Function]}
                        >
                          <button>
                            Contact
                          </button>
                        </a>
                      </Link>
                      <Link
                        replace={false}
                        to="/info"
                      >
                        <a
                          href="/info"
                          onClick={[Function]}
                        >
                          <button>
                            info
                          </button>
                        </a>
                      </Link>
                    </div>
                  </MainMenu>
                </header>
              </div>
            </Header>
            <div
              className="wrapper"
            >
              <div>
                <h3
                  data-testid="home-txt"
                >
                  Home Page
                </h3>
              </div>
              <p>
                <span>
                  <div>
                    Name: Abhishek
                  </div>
                  <div>
                    Location: Varanasi
                  </div>
                  <div>
                    Email: 
                  </div>
                </span>
              </p>
              <span>
                <input
                  data-testid="email-input"
                  name="email"
                  onChange={[Function]}
                  type="text"
                  value=""
                />
                <button
                  data-testid="button"
                  onClick={[Function]}
                >
                  Get language list
                </button>
              </span>
              <div>
                <p>
                  <span
                    data-testid="about-us-link"
                    onClick={[Function]}
                    style={
                      Object {
                        "cursor": "pointer",
                      }
                    }
                  >
                    About Us
                  </span>
                </p>
              </div>
            </div>
          </div>
        </Home>
      </Connect(Home)>
    </Router>
  </BrowserRouter>
</Provider>
`;
